### PR TITLE
Add ability to switch and remove user

### PR DESCRIPTION
**Pull request by leemaguire**
App now contains `switch_user(user_id)` & `remove_user(user_id)`

### CODE REVIEW

File: src/sync/app.hpp [hunk 1]
```diff
@@ -269,6 +269,23 @@ class App {
...
+    /// exist in the list of all users who have logged into this application, and
+    /// the user must be currently logged in, otherwise this will return an
+    /// AppError.
+    ///
+    /// @param user_id A String specifying the desired `user_id`
+    /// @param completion_block will return a SyncUser on success, an AppError on failure
+    void switch_user(const std::string& user_id,
+                     std::function<void(std::shared_ptr<SyncUser>, Optional<AppError>)> completion_block) const;
+    
+    /// Logs out and removes the user with the provided id
```
Comment (cmelchior) [inline src/sync/app.hpp]: Would probably be helpful to specify that this is only a local operation and doesn't affect the server in any way.

File: src/sync/app.hpp [hunk 2]
```diff
@@ -269,6 +269,23 @@ class App {
...
+    /// if no user ID is provided it will attempt to log out the current user and remove it
```
Comment (cmelchior) [inline src/sync/app.hpp]: It would probably be cleaner to have a public `remove_user` method that only accepts a completion block similar to how log_out works. This feels potentially confusing.

File: src/sync/sync_manager.cpp [hunk 3]
```diff
@@ -411,6 +411,32 @@ void SyncManager::set_current_user(const std::string& user_id)
...
+    
+    if (!m_metadata_manager) {
+        return;
+    }
+    
+    if (user->identity() == m_metadata_manager->get_current_user_identity()) {
+        m_metadata_manager->set_current_user_identity("");
+    }
+    
+    m_users.erase(it);
```
Comment (cmelchior) [inline src/sync/sync_manager.cpp]: This looks dangerous. SDK's might still have a pointer to the User after it has been removed. I suspect doing this will cause all operations to fail after this. See https://mongodb.slack.com/archives/CKW2HG3M0/p1584430859364900 for more context.

IMO it would make sense to rework the current User states, so instead of the current ones we have `LOGGED_IN, LOGGED_OUT and REMOVED. It seems to fit the semantics we offer in a better way.

File: tests/sync/app.cpp [hunk 4]
```diff
@@ -1190,3 +1181,173 @@ TEST_CASE("app: response error handling", "[sync][app]") {
...
+    reset_test_directory(base_path);
+
+    auto tsm = TestSyncManager(base_path);
+    
+    bool processed = false;
+    
+    std::string id_a;
+    std::string id_b;
+    
+    SECTION("remove user expect success") {
```
Comment (cmelchior) [inline tests/sync/app.cpp]: Missing tests of what happens if you access user after it was removed.

File: src/sync/app.hpp [hunk 5]
```diff
@@ -269,6 +269,23 @@ class App {
...
     void log_out(std::shared_ptr<SyncUser> user, std::function<void(Optional<AppError>)> completion_block) const;
+    
+    /// Switches the active user to the user with the specified id. The user must
+    /// exist in the list of all users who have logged into this application, and
+    /// the user must be currently logged in, otherwise this will return an
+    /// AppError.
+    ///
+    /// @param user_id A String specifying the desired `user_id`
+    /// @param completion_block will return a SyncUser on success, an AppError on failure
+    void switch_user(const std::string& user_id,
```
Comment (cmelchior) [inline src/sync/app.hpp]: Why use `user_id` instead of the accepting the `shared_ptr<SyncUser>`? If an SDK wants to Switch a user, the developer would already have a handle to that user which is in the form of a shared_ptr. This kind of API just seems to result in us needing to look things up more times?

File: src/sync/app.hpp (hunk 1 above)
Comment (leemaguire) [reply src/sync/app.hpp]: Makes sense, I've updated the doc to reflect that

File: src/sync/app.hpp (hunk 2 above)
Comment (leemaguire) [reply src/sync/app.hpp]: I've now added another method that just takes the completion block

File: tests/sync/app.cpp (hunk 4 above)
Comment (leemaguire) [reply tests/sync/app.cpp]: Added a check to make sure the `user_a` and `user_b` reflect the correct state after they are removed

File: src/sync/app.cpp [hunk 6]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
     log_out(current_user(), completion_block);
 }
 
+void App::switch_user(std::shared_ptr<SyncUser> user,
+                      std::function<void(std::shared_ptr<SyncUser>, Optional<AppError>)> completion_block) const
+{
+    int sync_user_idx = -1;
+    for(size_t i = 0; i < SyncManager::shared().all_users().size(); i++) {
```
Comment (jsflax) [inline src/sync/app.cpp]: [style] Simplify this to `auto it = std::find(SyncManager::shared().all_users().begin(), SyncManager::shared().all_users().end(), user)`. Then replace the below checks using the appropriate iterator comparisons, and read the user by dereferencing it from the iterator.

File: src/sync/app.cpp [hunk 7]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+    
+    if (sync_user_idx == -1) {
+        return completion_block(nullptr,
+                                AppError(make_custom_error_code(ClientErrorCode::user_not_found), "User does not exist"));
+    }
+    
+    auto sync_user = SyncManager::shared().all_users()[sync_user_idx];
+    if (!sync_user->is_logged_in()) {
+        return completion_block(nullptr,
+                                AppError(make_custom_error_code(ClientErrorCode::user_not_logged_in), "User is not longer valid"));
```
Comment (jsflax) [inline src/sync/app.cpp]: This is not entirely accurate, they may be logged out. Please add a clearer error message.

File: src/sync/app.cpp [hunk 8]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+    SyncManager::shared().set_current_user(user->identity());
+    
+    return completion_block(current_user(), {});
+}
+
+void App::remove_user(std::shared_ptr<SyncUser> user,
+                      std::function<void(Optional<AppError>)> completion_block) const
+{
+    int sync_user_idx = -1;
+    for(size_t i = 0; i < SyncManager::shared().all_users().size(); i++) {
```
Comment (jsflax) [inline src/sync/app.cpp]: [style] Same as above.

File: src/sync/app.cpp [hunk 9]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+        if (SyncManager::shared().all_users()[i] == user) {
+            sync_user_idx = static_cast<int>(i);
+            break;
+        }
+    }
+    
+    if (sync_user_idx == -1) {
```
Comment (jsflax) [inline src/sync/app.cpp]: [style] No need to if/else here, just if.

File: src/sync/app.cpp [hunk 10]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+        return completion_block(AppError(make_custom_error_code(ClientErrorCode::user_not_found),
+                                         "No user has been found"));
+    } else {
+        auto sync_user = SyncManager::shared().all_users()[sync_user_idx];
+        if (sync_user->is_logged_in()) {
```
Comment (jsflax) [inline src/sync/app.cpp]: [major] This needs to follow the full logout flow. You are only logging out the user locally.

File: src/sync/app.cpp [hunk 11]
```diff
@@ -622,5 +622,74 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+            sync_user->log_out();
+        }
+        SyncManager::shared().remove_user(user->identity());
+    }
+
+    return completion_block({});
+}
+
+void App::remove_user(std::function<void(Optional<AppError>)> completion_block) const
```
Comment (jsflax) [inline src/sync/app.cpp]: [q] Do we need this function?

File: src/sync/sync_manager.cpp [hunk 12]
```diff
@@ -411,6 +411,33 @@ void SyncManager::set_current_user(const std::string& user_id)
...
+    if (!m_metadata_manager) {
+        return;
+    }
+    
+    if (user->identity() == m_metadata_manager->get_current_user_identity()) {
+        m_metadata_manager->set_current_user_identity("");
+    }
+    
+    user->set_state(SyncUser::State::Removed);
+
```
Comment (jsflax) [inline src/sync/sync_manager.cpp]: [major] You also need to mark the user for removal. This is not fully removing the user.

Comment (leemaguire) [reply src/sync/sync_manager.cpp]: done

File: src/sync/app.cpp (hunk 11 above)
Comment (leemaguire) [reply src/sync/app.cpp]: It is potentially confusing as we do not know who we are removing, removed this function as a result of that

File: src/sync/app.cpp (hunk 10 above)
Comment (leemaguire) [reply src/sync/app.cpp]: done

File: src/sync/app.cpp (hunk 8 above)
Comment (leemaguire) [reply src/sync/app.cpp]: done

File: src/sync/app.cpp (hunk 9 above)
Comment (leemaguire) [reply src/sync/app.cpp]: done

File: src/sync/app.cpp (hunk 7 above)
Comment (leemaguire) [reply src/sync/app.cpp]: done

File: src/sync/app.cpp (hunk 6 above)
Comment (leemaguire) [reply src/sync/app.cpp]: done

File: src/sync/app.hpp (hunk 5 above)
Comment (leemaguire) [reply src/sync/app.hpp]: `shared_ptr<SyncUser>` is now the default parameter

File: src/sync/sync_manager.cpp (hunk 3 above)
Comment (leemaguire) [reply src/sync/sync_manager.cpp]: Changed, also I have updated the `SyncUser::State` enum to reflect your suggestion

**Comment by leemaguire** [comment]
@cmelchior can you PTAL

File: src/sync/app.cpp [hunk 13]
```diff
@@ -622,5 +622,53 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
     log_out(current_user(), completion_block);
 }
 
+void App::switch_user(std::shared_ptr<SyncUser> user,
```
Comment (cmelchior) [inline src/sync/app.cpp]: Why do we log the user out when switching between them? It should be possible to have multiple users logged in at the same time? This would also remove the need for the callback, which feels really awkward for this type of API.

File: tests/sync/app.cpp [hunk 14]
```diff
@@ -1190,3 +1181,244 @@ TEST_CASE("app: response error handling", "[sync][app]") {
...
+            CHECK(!error);
+            CHECK(SyncManager::shared().get_current_user() == user);
+            user_a = user;
+        });
+        
+        CHECK(user_a->state() == SyncUser::State::LoggedIn);
+        
+        app.log_out(user_a, [&](Optional<app::AppError> error) {
+            CHECK(!error);
+            // a logged out anon user will be marked as Error, not LoggedOut
```
Comment (cmelchior) [inline tests/sync/app.cpp]:
```suggestion
            // a logged out anon user will be marked as Removed, not LoggedOut
```

File: tests/sync/app.cpp [hunk 15]
```diff
@@ -1190,3 +1181,244 @@ TEST_CASE("app: response error handling", "[sync][app]") {
...
+            CHECK(user_a->state() == SyncUser::State::Removed);
+        });
+        
+        app.remove_user(user_a, [&](Optional<app::AppError> error) {
+            CHECK(error->message == "No user has been found");
```
Comment (cmelchior) [inline tests/sync/app.cpp]: Is there a reason that removing an already removed user is an error? Couldn't it just be a no-op. I don't have a strong preference though.

But the error message feels a bit weird. Maybe rephrase to "User has already been removed"?

File: src/sync/app.cpp [hunk 16]
```diff
@@ -622,5 +622,53 @@ void App::log_out(std::function<void (Optional<AppError>)> completion_block) con
...
+                        user);
+    
+    
+    if (it == SyncManager::shared().all_users().end()) {
+        return completion_block(AppError(make_custom_error_code(ClientErrorCode::user_not_found),
+                                         "No user has been found"));
+    }
+    
+    auto sync_user = *it;
+    SyncManager::shared().remove_user(sync_user->identity());
```
[the remark this answers is not in the document]
Comment (leemaguire) [reply src/sync/app.cpp]: True, I've added it into the remove into the log_out completion block, I've also added an else to handle if the user is logged out

File: src/sync/app.cpp (hunk 13 above)
Comment (leemaguire) [reply src/sync/app.cpp]: we don't `log_out` on `switch_user`, we only perform the log_out on `remove_user`

File: tests/sync/app.cpp (hunk 15 above)
Comment (leemaguire) [reply tests/sync/app.cpp]: I agree, a check has been added to see if they have already been removed

File: src/sync/app.cpp (hunk 13 above)
Comment (cmelchior) [reply src/sync/app.cpp]: Hmm, could have sworn there was code doing that, but why is there a callback with an Error in that case. That shouldn't be needed?

Comment (leemaguire) [reply src/sync/app.cpp]: If the user is not logged in an error will be returned in the callback

File: src/sync/app.cpp (hunk 6 above)
Comment (leemaguire) [reply src/sync/app.cpp]: using std::find seems to fail tests on linux, using the traditional for loop structure seems to solve the issues with the tests

File: src/sync/app.cpp (hunk 13 above)
Comment (cmelchior) [reply src/sync/app.cpp]: Yes, but why making it a callback based API? That just makes it more annoying for consumers. There is nothing in this method that seems to require it, so I would just throw it as an exception?

File: src/sync/sync_manager.cpp [hunk 17]
```diff
@@ -411,6 +411,34 @@ void SyncManager::set_current_user(const std::string& user_id)
...
+    if (it == m_users.end())
+        return;
+    
+    auto user = *it;
+    
+    if (!m_metadata_manager) {
+        return;
+    }
+    
+    for (size_t i = 0; i < m_metadata_manager->all_unmarked_users().size(); i++) {
```
Comment (jsflax) [inline src/sync/sync_manager.cpp]: [style] why not use `find_if` again?

File: src/sync/app.cpp (hunk 13 above)
Comment (leemaguire) [reply src/sync/app.cpp]: After talking offline, I've removed the callback and changed to returning a shared pointer to the user, and throwing an exception otherwise

File: src/sync/sync_manager.cpp (hunk 17 above)
Comment (leemaguire) [reply src/sync/sync_manager.cpp]: resolved offline
